macaroonbakery/macaroon.py

```diff
@@ -157,7 +167,38 @@ def _new_caveat_id(self, base):
         @param base string
         @return string
         '''
-        raise NotImplementedError
+        id = bytearray()
+        if len(base) > 0:
+            id.append(base)
+        else:
+            # Add a version byte to the caveat id.Technically
+            # this is unnecessary as the caveat-decoding logic
+            # that looks at versions should never see this id,
+            # but if the caveat payload isn't provided with the
+            # payload, having this version gives a strong indication
+            # that the payload has been omitted so we can produce
+            # a better error for the user.
+            id.append(BAKERY_V3)
+        id.append(0)
+        # Iterate through integers looking for one that isn't already used,
+        # starting from n so that if everyone is using this same algorithm,
+        # we'll only perform one iteration.
+        i = len(self._caveat_data)
+        caveats = self._macaroon.caveats
+        while True:
+            temp = bytearray()
```

Comment:
temp = id[:]
?

Reply:
fixed the whole algo in the next PR coming.